app/controllers/actions/IdentifierAction.scala

```diff
@@ -17,69 +17,180 @@
+        }
     } recover {
       case _: NoActiveSession =>
-        Redirect(config.loginUrl, Map("continue" -> Seq(config.loginContinueUrl)))
+        Redirect(appConfig.loginUrl, Map("continue" -> Seq(appConfig.loginContinueUrl)))
+
+      case _: InsufficientEnrolments =>
+        // Redirect to the non enrolled page; this is authed but doesn't need enrolments.
+        // There we can examine the user and determine where to send them.
+        Results.Redirect(homeRoutes.UnauthorisedController.notEnrolled())
+
       case _: AuthorisationException =>
-        Redirect(routes.UnauthorisedController.onPageLoad)
+        Redirect(homeRoutes.UnauthorisedController.unauthorised())
     }
   }
 
-}
-
-class SessionIdentifierAction @Inject() (val parser: BodyParsers.Default)(implicit
-  val executionContext: ExecutionContext
-) extends IdentifierAction {
+  private def authPredicate(selectedClientIdentifier: Option[String] = None): Predicate =
+    selectedClientIdentifier.map { clientIdentifier =>
+      // If this request is decorated with a selected client identifier this indicates
+      // an agent at work; we need to request the delegated authority
+      Enrolment(pptEnrolmentKey).withIdentifier(pptEnrolmentIdentifierName,
+                                                clientIdentifier
+      ).withDelegatedAuthRule("ppt-auth")
+    }.getOrElse {
+      Enrolment(pptEnrolmentKey)
+    }.and(acceptableCredentialStrength)
 
-  override def invokeBlock[A](
+  private def executeRequest[A](
     request: Request[A],
-    block: IdentifierRequest[A] => Future[Result]
-  ): Future[Result] = {
-
-    implicit val hc: HeaderCarrier =
-      HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+    block: IdentifiedRequest[A] => Future[Result],
+    identityData: IdentityData,
+    pptEnrolmentIdentifier: String,
+    allEnrolments: Enrolments
+  ) =
+    if (pptReferenceAllowedList.isAllowed(pptEnrolmentIdentifier)) {
+      val pptLoggedInUser = SignedInUser(allEnrolments, identityData)
+      block(new IdentifiedRequest(request, pptLoggedInUser, Some(pptEnrolmentIdentifier)))
+    } else {
+      logger.warn("User id is not allowed, access denied")
+      Future.successful(Results.Redirect(homeRoutes.UnauthorisedController.unauthorised()))
+    }
 
-    hc.sessionId match {
-      case Some(session) =>
-        block(IdentifierRequest(request, session.value))
-      case None =>
-        Future.successful(Redirect(routes.JourneyRecoveryController.onPageLoad()))
+  private def getPptEnrolmentId(
+    enrolments: Enrolments,
+    identifier: String,
+    selectedClientIdentifier: Option[String]
+  ): Option[String] =
+    // It appears Auth with never return a delegated enrolment in it's response to an agent auth request;
+    // therefore we have to use the one the agent past in. This is safe because this identifier has just
+    // past auth for this this agent.
+    selectedClientIdentifier.map(Some(_)).getOrElse {
```

Comment:
this feels a bit convoluted, unwrapping options to rewrap them, think this can be refactored, happy to help